### PR TITLE
caddy: disable telemetry

### PR DESCRIPTION
**Pull request by marsam**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by kalbasit** [comment]
What's the reason for disabling telemetry?

**Comment by gloaming** [comment]
@kalbasit Do we not disable telemetry as a matter of policy? It's disabled in firefox, and searching the repo I see that it's disabled in Powershell and TimescaleDB also.

File: pkgs/servers/caddy/default.nix [hunk 1]
```diff
@@ -16,9 +16,13 @@ buildGoModule rec {
...
-  buildFlagsArray = ''
-    -ldflags=
-      -X github.com/mholt/caddy/caddy/caddymain.gitTag=v${version}
+  buildFlagsArray = [
+    "-ldflags=" "-w -s -X github.com/mholt/caddy/caddy/caddymain.gitTag=v${version}"
+  ];
+
+  preBuild = ''
+    substituteInPlace caddy/caddymain/run.go \
+      --replace "var EnableTelemetry = true" "var EnableTelemetry = false"
```
Comment (Mic92) [inline pkgs/servers/caddy/default.nix]: Can you make this a patch instead?

**Comment by Mic92** [comment]
Making a software not transmitting information to third-party services by default is a good practice, we follow in NixOS.

File: pkgs/servers/caddy/default.nix (hunk 1 above)
Comment (kalbasit) [reply pkgs/servers/caddy/default.nix]: also nitpick: `var EnableTelemetry = false` is the same as `var EnableTelemetry bool`.

**Comment by fpletz** [comment]
The recommended way by upstream to disable telemetry is to create an new main module: https://github.com/caddyserver/caddy#build. Since we also need this to package caddy modules in the future, I've rewritten this PR to overwrite the default main module with a custom one to disable the telemetry. Also rebased onto the current master.

You can check if telemetry is disabled by checking if `caddy -disabled-metrics foo` is failing with an error.

**Comment by marsam** [comment]
@fpletz it works! thanks a lot for updating it

**Pull request merged**